Adding tests for IsCategorical Method and Refactoring GithubLabel Methods

seanlowjk opened the pull request:
# Summary 

Thus PR fixes #467, which adds unit tests for the `isCategorical` of `GithubLabel` 

## Description 

Some things to note: 

1. Testing with mutliple dots returns an unexpected value for `getValue`

For example:

```
githubLabel = new GithubLabel({ name: `${category}.....${value}` });
githubLabel.getLabel(); // returns " "
```

Not sure if this is the behavior we would want! Just needa double check here!

anubh-v commented:
Thanks @seanlowjk 
We can make the regex stricter so that it only considers strings of the form `category.value` as categorical (instead of allowing any non-zero number of periods).

I also have a suggestion to refactor the `GitHubLabel` class:
Perform the `isCategorical` regex test and extract the category name and value just once within the `GitHubLabel` constructor, instead of performing the regex test everytime `getValue` and `getCategory` are called.

@seanlowjk  @JunWei96 what do you think?
We could do the refactor in this PR, since we're adding the relevant tests.

seanlowjk commented:
> Thanks @seanlowjk
> The tests look good.
> 
> We can make the regex stricter so that it only considers strings of the form `category.value` as categorical (instead of allowing any non-zero number of periods). This will be better as it gives us fewer surprises.
> 
> I also have a suggestion to refactor the `GitHubLabel` class:
> Perform the `isCategorical` regex test and extract the category name and value just once within the `GitHubLabel` constructor, instead of performing the regex test everytime `getValue` and `getCategory` are called.
> 
> @seanlowjk @JunWei96 what do you think?
> We could do the refactor in this PR, since we're adding the relevant tests.

Yea, it'll be good to do so! Then can reduce the number of calls will be good

seanlowjk commented:
Just a note, I have updated the constructor to do the checks for `isCategorical` only **in the constructor** 

As a result, we would only need 1 call for each and every GithubLabel. 

```
  constructor(githubLabels: {}) {
    Object.assign(this, githubLabels);

    if (this.isCategorical()) {
      this.category = this.name.split('.')[0];
      this.label = this.name.split('.')[1];
    } else {
      this.category = this.name;
      this.label = this.name;
    }
    Object.freeze(this);
  }
```

Not sure if this is the best way to refactor the methods, so just leaving it open for discussion first!

anubh-v commented on tests/app/core/models/github/github-label.model.spec.ts:
```diff
@@ -0,0 +1,41 @@
+  const category = 'severity';
+  const value = 'LOW';
+
+  it('should returns false on isCategorical() when provided wit no name', () => {
```
Minor nitpick: let's phrase these descriptions as `.isCategorical() should return <bool> when ....`

seanlowjk replied:
Eyes! :+1:

anubh-v commented on tests/app/core/models/github/github-label.model.spec.ts:
```diff
@@ -0,0 +1,41 @@
+    githubLabel = new GithubLabel({ name: `${category}.${value}` });
+    expect(githubLabel.isCategorical()).toEqual(true);
+
+    githubLabel = new GithubLabel({ name: `${category}.......${value}` });
```
After the recent regex change, this case (and the following case) should not be grouped with the positive tests right?

seanlowjk replied:
Yes Yes, i will split it into the relevant groups!

anubh-v commented on tests/app/core/models/github/github-label.model.spec.ts:
```diff
@@ -0,0 +1,41 @@
+    expect(githubLabel.isCategorical()).toEqual(false);
+  });
+
+  it('should returns true on isCategorical() when provided with valid label name', () => {
```
Let's change the spec description to `... when provided with categorical label name`.

seanlowjk replied:
:+1:

anubh-v replied:
I think you missed this

seanlowjk replied:
Okays!

seanlowjk replied:
Should have fixed the tests naming convention already!

anubh-v reviewed (commented):
Looks good to me, besides some minor formatting & phrasing suggestions.
